### ISSUE TITLE
dill load and sklearn clone result in error

### ISSUE THREAD

**Issue by DCoupry** [skorch-dev/skorch#1026]
Dumping a skorch model with dill then reloading it (does not matter if with dill or pickle) makes it incompatible with sklearn.base.clone , apparently due to some attributes becoming empty - the optimizers I think, but Ihad no time to investigate further. This behaviour occurs neither with pickle or joblib.

This makes using functions such as cross_val_predict unuseable after loading a previously dumped model.

to reproduce:

python=3.10
tried with a bunch of versions for dill / torch / skorch / sklearn. all bug out.

`
from sklearn.datasets import make_regression
from sklearn.base import clone
import torch
import skorch
import dill

X, y = make_regression()
base_model = skorch.NeuralNetRegressor(torch.nn.Linear(100,1))
cloned_model = clone(base_model)
dumped_model = dill.loads(dill.dumps(base_model))
cloned_dumped_model = clone(dumped_model)

base_model.fit(X, y) # works
cloned_model .fit(X, y) # works
dumped_model .fit(X, y) # works
cloned_dumped_model .fit(X, y) # does not work
`

**Comment by DCoupry** [comment]
I can reproduce it, yes. And indeed the dumped version does die also. What does work is:

```python
dumped_model = dill.loads(pickle.dumps(base_model))
dumped_model.fit(X, y)
```

the error is the same. the loss is None here. The process looks okay to me and goes through all the initializations, and I have tracked it to the train_step function, where if you print the optimizers you will get an empty list. But when you take the models themselves, and print the pre-fit atributes, everything looks good! quite frustrating.

**Comment by DCoupry** [comment]
okay, after some checks:

```python
from sklearn.datasets import make_regression
from sklearn.base import clone
import numpy as np
import torch
import skorch
import dill
import pickle

dill.__version__  # 0.3.6

X, y = make_regression()
X, y = X.astype(np.float32), y.astype(np.float32).reshape(-1, 1)  # added
base_model = skorch.NeuralNetRegressor(torch.nn.Linear(100, 1))
cloned_model = clone(base_model)
dumped_model = dill.loads(dill.dumps(base_model))
dumped_fitted_model = dill.loads(dill.dumps(base_model.fit(X, y)))
cloned_dumped_model = clone(dumped_model)
cloned_dumped_fitted_model = clone(cloned_dumped_model)

base_model.fit(X, y) # works
cloned_model.fit(X, y) # works
dumped_model.fit(X, y) # fails
dumped_fitted_model.fit(X, y) # works
cloned_dumped_fitted_model.fit(X, y) # fails
cloned_dumped_model.fit(X, y) # fails with same error
```

**Comment by BenjaminBossan** [comment]
Thanks for investigating further. This is super strange IMO, because the `_optimizers` attribute is empty but `_modules` and `_crtiteria` are not empty, even though these three attributes are treated exactly the same. Do you know if dill uses `__getstate__` and `__setstate__` or if it has equivalent methods? Maybe we can salvage something there.

**Comment by DCoupry** [comment]
we could print a trace of execution with the final fit and diff across, maybe?

**Comment by BenjaminBossan** [comment]
Sorry, I don't understand. How can this be done?

**Comment by DCoupry** [comment]
After trying unsuccessfully to solve this problem (might revisit in a while, but no promises), I suggest we add a caveat in https://skorch.readthedocs.io/en/stable/user/save_load.html and close the issue?

**Comment by BenjaminBossan** [comment]
SGTM, it could mention the workaround you found above. Are you interested in creating a PR for it?

**Comment by DCoupry** [comment]
will try and make one soonish.
